File: python/bricker/game_stats.py
"""Bricker - A Tetris-like brick game.
Copyright (C) 2017-2020  John Hyland
GNU GENERAL PUBLIC LICENSE Version 3"""

from typing import List
import sys
import os.path


class GameStats:
    """Stores current score, high scores, and other game statistics."""

    def __init__(self) -> None:
        """Class constructor."""
        self.__high_scores: List[HighScore] = self.__load_high_scores()
        self.__current_score: int = 0
        self.__lines: int = 0
        self.__level: int = 1

    @property
    def high_scores(self) -> List['HighScore']:
        """Returns list of high scores."""
        return self.__high_scores

    @property
    def current_score(self) -> int:
        """Returns the current score."""
        return self.__current_score

    @property
    def lines(self) -> int:
        """Returns the number of lines cleares."""
        return self.__lines

    @property
    def level(self) -> int:
        """Returns the current level."""
        return self.__level

    @level.setter
    def level(self, value: int) -> None:
        """Sets the current level."""
        self.__level = value

    def __load_high_scores(self) -> List['HighScore']:
        """Load high scores from file."""
        scores = []
        if os.path.isfile("high_scores.txt"):
            with open("high_scores.txt", "r") as f:
                lines = f.readlines()
            lines = [x.strip() for x in lines]
            for line in lines:
                split = line.split("\t")
                if len(split) == 2:
                    initials = split[0]
                    score = int(split[1])
                    scores.append(HighScore(initials, score))
            scores = self.__sort_scores(scores)
        return scores

    def __save_high_scores(self, scores: List['HighScore']) -> None:
        """Save high scores to file."""
        scores = self.__sort_scores(scores)
        with open("high_scores.txt", "w") as text_file:
            for x in scores:
                text_file.write(x.initials + "\t" + str(x.score) + "\n")

    def is_high_score(self) -> bool:
        """Returns true if score can be placed on board."""
        if len(self.__high_scores) < 10:
            return True
        lowest = sys.maxsize
        for score in self.__high_scores:
            if score.score < lowest:
                lowest = score.score
        return self.__current_score > lowest

    def add_high_score(self, initials: str) -> None:
        """Adds new score, sorts and limits to top 10, saves to disk."""
        self.__high_scores.append(HighScore(initials, self.__current_score))
        self.__high_scores = self.__sort_scores(self.__high_scores)
        self.__save_high_scores(self.__high_scores)

    @staticmethod
    def __sort_scores(scores: List['HighScore']) -> List['HighScore']:
        """Sorts scores and returns new list.  Truncates to top ten."""
        scores.sort(key=lambda x: x.score, reverse=True)
        while len(scores) > 10:
            del scores[-1]
        return scores

    def add_lines(self, count: int) -> None:
        """Increments cleared lines, sets level."""
        self.__lines += count
        self.__level = (self.__lines // 20) + 1

    def increment_score(self, value: int) -> None:
        """Increments current score by specified value."""
        self.__current_score += value


class HighScore:
    """Stores a single high score."""

    def __init__(self, initials: str, score: int) -> None:
        """Class constructor."""
        self.__initials: str = initials
        self.__score: int = score

    @property
    def initials(self) -> str:
        """Returns gamer's initials."""
        return self.__initials

    @property
    def score(self) -> int:
        """Returns high-score value."""
        return self.__score

File: csharp/Bricker/Game/GameStats.cs
using Bricker.Configuration;
using Bricker.Error;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace Bricker.Game
{
    /// <summary>
    /// Stores game statistics.
    /// </summary>
    public class GameStats
    {
        //private
        private readonly Config _config;
        private List<HighScore> _highScores;
        private int _score;
        private int _lines;
        private int _level;
        private int _linesSent;
        private bool _gameOver;

        //public
        public IReadOnlyList<HighScore> HighScores => _highScores.AsReadOnly();
        public int Score => _score;
        public int Lines => _lines;
        public int Level => _level;
        public int LinesSent => _linesSent;
        public bool GameOver => _gameOver;

        /// <summary>
        /// Class constructor.
        /// </summary>
        public GameStats(Config config)
        {
            _config = config;
            _highScores = LoadHighScores();
            _score = 0;
            _lines = 0;
            _level = 1;
            _linesSent = 0;
            _gameOver = false;
        }

        /// <summary>
        /// Resets stats for new game.
        /// </summary>
        public void Reset()
        {
            _highScores = LoadHighScores();
            _score = 0;
            _lines = 0;
            _level = 1;
            _linesSent = 0;
            _gameOver = false;
        }

        /// <summary>
        /// Increments current score.
        /// </summary>
        public void IncrementScore(int value)
        {
            _score += value;
        }

        /// <summary>
        /// Increments cleared line count, and sets level.
        /// </summary>
        public void IncrementLines(int value)
        {
            _lines += value;
            _level = (_lines / 25) + 1;
        }

        /// <summary>
        /// Sets the current level.
        /// </summary>
        public void SetLevel(int level)
        {
            if (level > 10)
                level = 10;
            if (level < 1)
                level = 1;
            _level = level;
        }

        /// <summary>
        /// Increments lines sent (two-player mode).
        /// </summary>
        public void IncrementLinesSent(int value)
        {
            _linesSent += value;
        }

        /// <summary>
        /// Sets game-over flag to true (means local player lost).
        /// </summary>
        public void SetGameOver()
        {
            _gameOver = true;
        }

        /// <summary>
        /// Returns true if score can be placed on board.
        /// </summary>
        public bool IsHighScore()
        {
            if (_highScores.Count < 10)
                return true;
            int lowest = Int32.MaxValue;
            foreach (HighScore score in _highScores)
                if (score.Score < lowest)
                    lowest = score.Score;
            return _score > lowest;
        }

        /// <summary>
        /// Adds a new high score, sorts and keeps top 10, saves to disk.
        /// </summary>
        public void AddHighScore(string initials)
        {
            _highScores.Add(new HighScore(initials, _score));
            _highScores = _highScores.OrderByDescending(s => s.Score).Take(10).ToList();
            SaveHighScores();
        }

        /// <summary>
        /// Saves high scores to disk.
        /// </summary>
        private void SaveHighScores()
        {
            try
            {
                string text = String.Join(Environment.NewLine, _highScores.Select(s => $"{s.Initials}\t{s.Score}"));
                File.WriteAllText(_config.HighScoreFile, text);
            }
            catch (Exception ex)
            {
                ErrorHandler.LogError(ex);
            }
        }

        /// <summary>
        /// Returns list of high scores from disk, or empty list if file not found.
        /// </summary>
        private List<HighScore> LoadHighScores()
        {
            try
            {
                if (!File.Exists(_config.HighScoreFile))
                    return new List<HighScore>();
                string[] lines = File.ReadAllLines(_config.HighScoreFile);
                List<HighScore> scores = new List<HighScore>();
                foreach (string line in lines)
                {
                    string[] split = line.Trim().ToUpper().Split('\t');
                    if (split.Length != 2)
                        continue;
                    try
                    {
                        string initials = Config.CleanInitials(split[0]);
                        int score = Int32.Parse(split[1]);
                        scores.Add(new HighScore(initials, score));
                    }
                    catch (Exception ex)
                    {
                        ErrorHandler.LogError(ex);
                    }
                }
                return scores.OrderByDescending(s => s.Score).Take(10).ToList();
            }
            catch (Exception ex)
            {
                ErrorHandler.LogError(ex);
            }
            return new List<HighScore>();
        }



    }

    /// <summary>
    /// Represents a single high score.
    /// </summary>
    public class HighScore
    {
        public string Initials { get; }
        public int Score { get; }

        public HighScore(string initials, int score)
        {
            Initials = initials;
            Score = score;
        }
    }
}
